ion/textutils.go

```diff
@@ -342,39 +342,97 @@ func parseTimestamp(val string) (time.Time, error) {
 		return invalidTimestamp(val)
 	}
 
 	if len(val) == 10 || (len(val) == 11 && (val[10] == 't' || val[10] == 'T')) {
 		// yyyy-mm-dd or yyyy-mm-ddT
-		return time.Date(int(year), time.Month(month), int(day), 0, 0, 0, 0, time.UTC), nil
+		return tryCreateTimeDate(val, year, month, day, time.UTC)
 	}
 	if val[10] != 't' && val[10] != 'T' {
 		return invalidTimestamp(val)
 	}
 
+	// At this point timestamp must have hour:minute
 	if len(val) < 17 {
 		return invalidTimestamp(val)
 	}
-	if val[16] != ':' {
-		return time.Parse("2006-01-02T15:04Z07:00", val)
+	if val[16] == 'z' || val[16] == 'Z' {
+		return time.Parse("2006-01-02T15:04Z", val)
+	}
+	if val[16] == '+' || val[16] == '-' {
+		if isValidOffset(val, 16) {
+			// TODO: we should use time.ParseInLocation()
+			return time.Parse("2006-01-02T15:04Z07:00", val)
+		}
+		return invalidTimestamp(val)
 	}
+	if val[16] == ':' {
+		//yyyy-mm-ddThh:mm:ss
+		if len(val) < 20 {
+			return invalidTimestamp(val)
+		}
+
+		idx := 19
+		if val[idx] == '.' {
+			idx++
+			for idx < len(val) && isDigit(int(val[idx])) {
+				idx++
+			}
+		}
 
-	if len(val) > 19 && val[19] == '.' {
-		i := 20
-		for i < len(val) && isDigit(int(val[i])) {
-			i++
+		if val[idx] == 'z' || val[idx] == 'Z' {
+			if idx >= 29 {
+				// Too much precision for a go Time.
+				// TODO: We should probably round instead of truncating? Ah well.
+				return time.Parse(time.RFC3339Nano, val[:29]+val[idx:])
+			}
+			return time.Parse(time.RFC3339, val)
 		}
 
-		if i >= 29 {
-			// Too much precision for a go Time.
-			// TODO: We should probably round instead of truncating? Ah well.
-			return time.Parse(time.RFC3339Nano, val[:29]+val[i:])
+		if val[idx] == '+' || val[idx] == '-' {
+			if isValidOffset(val, idx) {
+				if idx >= 29 {
+					// Too much precision for a go Time.
+					// TODO: We should probably round instead of truncating? Ah well.
+					// TODO: we should use time.ParseInLocation()
+					return time.Parse(time.RFC3339Nano, val[:29]+val[idx:])
+				}
+				// TODO: we should use time.ParseInLocation()
+				return time.Parse(time.RFC3339, val)
+			}
+			return invalidTimestamp(val)
 		}
 	}
+	return invalidTimestamp(val)
+}
+
+func tryCreateTimeDate(val string, year int64, month int64, day int64, loc *time.Location) (time.Time, error) {
+	date := time.Date(int(year), time.Month(month), int(day), 0, 0, 0, 0, loc)
+	// time.Date converts 2000-01-32 input to 2000-02-01
+	if int(year) != date.Year() || time.Month(month) != date.Month() || int(day) != date.Day() {
+		return invalidTimestamp(val)
+	}
+	return date, nil
+}
+
+func isValidOffset(val string, idx int) bool {
+	// +hh:mm
+	if idx+5 > len(val) || val[idx+3] != ':' {
+		return false
+	}
+
+	hourOffset, errH := strconv.ParseInt(val[idx+1:idx+3], 10, 32)
+	minuteOffset, errM := strconv.ParseInt(val[idx+4:], 10, 32)
+	if errH != nil || errM != nil {
+		return false
+	}
+
+	// TODO: check the offset against max offset (14 * 60), adjust and return if larger than max
```

Comment:
Per our earlier conversation, I don't think we should change the provided offset unless RFC3339 (on which `timestamp` is based) says that there's a max.